### PR TITLE
Update FFG Check Points During Block Processing 

### PR DESCRIPTION
**Pull request by terenc3t**
**Note**: Do not merge, still working on debugging run time with this PR 

How we saved finalized state after fork choice was wrong. We should be saving FFG check points (last finalized state, block and last justified state, block) after state execution so fork choice can use those information in the beginning.

Key changes:
- Implemented `updateFFGCheckPts` which saves finalized and justified objects
- Tests for `updateFFGCheckPts`
- Run `updateFFGCheckPts` after state execution versus after fork choice

### CODE REVIEW

File: beacon-chain/blockchain/stategenerator/state_generator.go [hunk 1]
```diff
@@ -21,6 +22,13 @@ func GenerateStateFromBlock(ctx context.Context, db *db.BeaconDB, block *pb.Beac
 	if err != nil {
 		return nil, err
 	}
+	log.WithFields(logrus.Fields{
+		"input_block_slot":  block.Slot - params.BeaconConfig().GenesisSlot,
+		"output_state_slot": hState.Slot - params.BeaconConfig().GenesisSlot,
```
Comment (rauljordan) [inline beacon-chain/blockchain/stategenerator/state_generator.go]: change these log fields to outputStateSlot (camelCase format)

Comment (terenc3t) [reply beacon-chain/blockchain/stategenerator/state_generator.go]: oops I need to delete this, it's mainly for debugging, it serves no purpose in production...

**Comment by prestonvanloon** [comment]
This doesn’t get beyond slot 79 in multinode deployment

File: beacon-chain/db/state.go [hunk 2]
```diff
@@ -173,12 +172,11 @@ func (db *BeaconDB) SaveFinalizedState(beaconState *pb.BeaconState) error {
 
 // SaveHistoricalState saves the last finalized state in the db.
 func (db *BeaconDB) SaveHistoricalState(beaconState *pb.BeaconState) error {
-	finalizedSlot := beaconState.FinalizedEpoch * params.BeaconConfig().SlotsPerEpoch
-	slotDiff := beaconState.Slot - finalizedSlot
+	slotDiff := beaconState.Slot - params.BeaconConfig().GenesisSlot
 
 	// Do not save state, if slot diff is not
```
Comment (prestonvanloon) [inline beacon-chain/db/state.go]: Update comment?

File: beacon-chain/db/state.go [hunk 3]
```diff
@@ -173,12 +172,11 @@ func (db *BeaconDB) SaveFinalizedState(beaconState *pb.BeaconState) error {
...
 	// a power of 2.
-	if !mathutil.IsPowerOf2(slotDiff) {
+	if slotDiff%params.BeaconConfig().SlotsPerEpoch != 0 {
```
Comment (prestonvanloon) [inline beacon-chain/db/state.go]: Why is this different?

File: beacon-chain/db/state.go [hunk 4]
```diff
@@ -275,9 +273,10 @@ func (db *BeaconDB) HistoricalStateFromSlot(slot uint64) (*pb.BeaconState, error
 	if err != nil {
 		return nil, fmt.Errorf("unable to retrieve finalized state %v", err)
 	}
-	finalizedSlot := state.FinalizedEpoch * params.BeaconConfig().SlotsPerEpoch
-	slotDiff := slot - finalizedSlot
-	closestSlot := mathutil.ClosestPowerOf2(slotDiff)
+	slotDiff := slot - params.BeaconConfig().GenesisSlot
+	if slotDiff%params.BeaconConfig().SlotsPerEpoch != 0 {
```
Comment (prestonvanloon) [inline beacon-chain/db/state.go]: Why?

**Pull request merged**